From b6929f1744fd4339338cae0cfb1905c560cb3443 Mon Sep 17 00:00:00 2001
From: Andreas Scherer <andreas_github@freenet.de>
Date: Mon, 18 May 2026 15:48:14 +0200
Subject: [PATCH 48/73] Prepare SGB for CPLUSPLUS23.

Compile with Apple clang 21.0.0 and --std=c++23.

* sprintf is deprecated: Replace with snprintf.
  Be extra carefull with appended material (pointers p and q)!
* register is not allowed in C++17.
---
 PROTOTYPES/gb_basic.ch | 444 ++++++++++++++++++++++++++++++++++++++++-
 PROTOTYPES/gb_flip.ch  |  10 +
 PROTOTYPES/gb_graph.ch |  72 +++++--
 PROTOTYPES/gb_io.ch    |  42 +++-
 PROTOTYPES/gb_sort.ch  |   6 +
 5 files changed, 546 insertions(+), 28 deletions(-)

diff --git a/PROTOTYPES/gb_basic.ch b/PROTOTYPES/gb_basic.ch
index ab8d35d..857ffcc 100644
--- a/PROTOTYPES/gb_basic.ch
+++ b/PROTOTYPES/gb_basic.ch
@@ -82,12 +82,48 @@ Graph *board(
   @<Insert arcs or edges for all legal moves@>@;
 @z
 
+@x [9] l.200
+register long i,j,k; /* all-purpose indices */
+register long d; /* the number of dimensions */
+register Vertex *v; /* the current vertex of interest */
+register long s; /* accumulator */
+@y
+long i,j,k; /* all-purpose indices */
+long d; /* the number of dimensions */
+Vertex *v; /* the current vertex of interest */
+long s; /* accumulator */
+@z
+
 @x [13] l.265
 @<Set up a graph with |n| vertices@>=
 @y
 @<Set up a graph with \9{n}|n| vertices@>=
 @z
 
+@x [13] l.275
+  sprintf(new_graph->id,"board(%ld,%ld,%ld,%ld,%ld,%ld,%d)",
+@y
+  snprintf(new_graph->id,ID_FIELD_SIZE,"board(%ld,%ld,%ld,%ld,%ld,%ld,%d)",
+@z
+
+@x [14] l.295
+{@+register char *q; /* string pointer */
+@y
+{@+char *q; /* string pointer */
+@z
+
+@x [14] l.301
+      sprintf(q,".%ld",xx[k]);
+@y
+      snprintf(q,BUF_SIZE-(q-buffer),".%ld",xx[k]);
+@z
+
+@x [16] l.342
+{@+register long w=wrap;
+@y
+{@+long w=wrap;
+@z
+
 @x [22] l.410
     do yy[k]+=nn[k];@+ while (yy[k]<0);
 @y
@@ -142,6 +178,18 @@ Graph *simplex(
         else {@+if (n4>(long)n) n4=n;
 @z
 
+@x [28] l.541
+sprintf(new_graph->id,"simplex(%lu,%ld,%ld,%ld,%ld,%ld,%d)",
+@y
+snprintf(new_graph->id,ID_FIELD_SIZE,"simplex(%lu,%ld,%ld,%ld,%ld,%ld,%d)",
+@z
+
+@x [29] l.551
+  register long *coef=gb_typed_alloc(n+1,long,working_storage);
+@y
+  long *coef=gb_typed_alloc(n+1,long,working_storage);
+@z
+
 @x [30] l.575
   for (k=1;k<=n;k++) {
 @y
@@ -156,6 +204,36 @@ if (yy[0]>=(long)n) {
   k=0;@+xx[0]=(yy[1]>=(long)n? 0: n-yy[1]);
 @z
 
+@x [34] l.647
+{@+register char *p=buffer; /* string pointer */
+@y
+{@+char *p=buffer; /* string pointer */
+@z
+
+@x [34] l.649
+    sprintf(p,".%ld",xx[k]);
+@y
+    snprintf(p,BUF_SIZE-(p-buffer),".%ld",xx[k]);
+@z
+
+@x [35] l.663
+  if (xx[j]) {@+register Vertex *u; /* previous vertex adjacent to |v| */
+@y
+  if (xx[j]) {@+Vertex *u; /* previous vertex adjacent to |v| */
+@z
+
+@x [35] l.666
+      if (xx[k]<nn[k]) {@+register char *p=buffer; /* string pointer */
+@y
+      if (xx[k]<nn[k]) {@+char *p=buffer; /* string pointer */
+@z
+
+@x [35] l.669
+          sprintf(p,".%ld",xx[i]);
+@y
+          snprintf(p,BUF_SIZE-(p-buffer),".%ld",xx[i]);
+@z
+
 @x [37] l.733
 Graph *subsets(n,n0,n1,n2,n3,n4,size_bits,directed)
   unsigned long n; /* the number of elements in the multiset */
@@ -170,6 +248,12 @@ Graph *subsets(
   long directed) /* should the graph be directed? */
 @z
 
+@x [38] l.752
+sprintf(new_graph->id,"subsets(%lu,%ld,%ld,%ld,%ld,%ld,0x%lx,%d)",
+@y
+snprintf(new_graph->id,ID_FIELD_SIZE,"subsets(%lu,%ld,%ld,%ld,%ld,%ld,0x%lx,%d)",
+@z
+
 @x [39] l.763
 if (yy[0]>=n) {
   k=0;@+xx[0]=(yy[1]>=n? 0: n-yy[1]);
@@ -178,9 +262,13 @@ if (yy[0]>=(long)n) {
   k=0;@+xx[0]=(yy[1]>=(long)n? 0: n-yy[1]);
 @z
 
-@x [40] l.790
+@x [40] l.788
+{@+register Vertex *u;
+  for (u=new_graph->vertices;u<=v;u++) {@+register char *p=u->name;
     long ss=0; /* the number of elements common to |u| and |v| */
 @y
+{@+Vertex *u;
+  for (u=new_graph->vertices;u<=v;u++) {@+char *p=u->name;
     unsigned long ss=0; /* the number of elements common to |u| and |v| */
 @z
 
@@ -196,6 +284,12 @@ Graph *perms(
   long directed) /* should the graph be directed? */
 @z
 
+@x [43] l.892
+  register long n; /* total number of elements in multiset */
+@y
+  long n; /* total number of elements in multiset */
+@z
+
 @x [44] l.908
 @<Determine |n| and the maximum possible number of inversions@>;
 @y
@@ -211,7 +305,13 @@ Graph *perms(
 @x [45] l.915
 {@+register long ss; /* max inversions known to be possible */
 @y
-{@+register unsigned long ss; /* max inversions known to be possible */
+{@+unsigned long ss; /* max inversions known to be possible */
+@z
+
+@x [46] l.934
+  register long *coef=gb_typed_alloc(max_inv+1,long,working_storage);
+@y
+  long *coef=gb_typed_alloc(max_inv+1,long,working_storage);
 @z
 
 @x [46] l.940
@@ -220,12 +320,30 @@ Graph *perms(
   for (k=1,nverts=1;k<=(long)max_inv;k++) {
 @z
 
+@x [46] l.948
+  sprintf(new_graph->id,"perms(%ld,%ld,%ld,%ld,%ld,%lu,%d)",
+@y
+  snprintf(new_graph->id,ID_FIELD_SIZE,"perms(%ld,%ld,%ld,%ld,%ld,%lu,%d)",
+@z
+
+@x [47] l.959
+for (k=1;k<=nn[j];k++) {@+register long ii;
+@y
+for (k=1;k<=nn[j];k++) {@+long ii;
+@z
+
 @x [47] l.961
   for (i=k,ii=0;i<=max_inv;i++,ii++) {
 @y
   for (i=k,ii=0;i<=(long)max_inv;i++,ii++) {
 @z
 
+@x [48] l.981
+{@+register long *xtab,*ytab,*ztab; /* permutations and their inversions */
+@y
+{@+long *xtab,*ytab,*ztab; /* permutations and their inversions */
+@z
+
 @x [50] l.1017
   if (m<max_inv && ytab[k]<k-1)
 @y
@@ -236,13 +354,25 @@ Graph *perms(
 static char *short_imap="0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ\
 abcdefghijklmnopqrstuvwxyz_^~&@@,;.:?!%#$+-*/|<=>()[]{}`'";
 @y
-static char *short_imap=
+static const char *short_imap=
   "0123456789"@|
   "ABCDEFGHIJKLMNOPQRSTUVWXYZ"@|
   "abcdefghijklmnopqrstuvwxyz"@|
   "_^~&@@,;.:?!%#$+-*/|<=>()[]{}`'";
 @z
 
+@x [52] l.1042
+{@+register char *p; register long *q;
+@y
+{@+char *p; long *q;
+@z
+
+@x [53] l.1056
+  if (xtab[j]>xtab[j+1]) {@+register Vertex *u;
+@y
+  if (xtab[j]>xtab[j+1]) {@+Vertex *u;
+@z
+
 @x [55] l.1099
 Graph *parts(n,max_parts,max_size,directed)
   unsigned long n; /* the number being partitioned */
@@ -257,6 +387,12 @@ Graph *parts(
   long directed) /* should the graph be directed? */
 @z
 
+@x [56] l.1124
+  register long *coef=gb_typed_alloc(n+1,long,working_storage);
+@y
+  long *coef=gb_typed_alloc(n+1,long,working_storage);
+@z
+
 @x [56] l.1127
   for (k=1;k<=max_parts;k++) {
 @y
@@ -269,12 +405,44 @@ Graph *parts(
     for (j=k,i=0;j<=(long)n;i++,j++) {
 @z
 
+@x [56] l.1139
+  sprintf(new_graph->id,"parts(%lu,%lu,%lu,%d)",
+@y
+  snprintf(new_graph->id,ID_FIELD_SIZE,"parts(%lu,%lu,%lu,%d)",
+@z
+
+@x [60] l.1190
+{@+register char *p=buffer; /* string pointer */
+@y
+{@+char *p=buffer; /* string pointer */
+@z
+
+@x [60] l.1192
+    sprintf(p,"+%ld",xx[k]);
+@y
+    snprintf(p,BUF_SIZE-(p-buffer),"+%ld",xx[k]);
+@z
+
 @x [61] l.1207
 if (d<max_parts) {
 @y
 if (d<(long)max_parts) {
 @z
 
+@x [62] l.1226
+{@+register Vertex *u; /* previous vertex adjacent to |v| */
+  register char *p=buffer;
+@y
+{@+Vertex *u; /* previous vertex adjacent to |v| */
+  char *p=buffer;
+@z
+
+@x [62] l.1234
+    sprintf(p,"+%ld",nn[k]);
+@y
+    snprintf(p,BUF_SIZE-(p-buffer),"+%ld",nn[k]);
+@z
+
 @x [63] l.1277
 The partial ordering corresponding to the directed graph on
 ${2n+1\choose n}/(2n+1)$ vertices created by |all_trees(n,1)|
@@ -305,18 +473,50 @@ Graph *binary(
     for (j=2;j<=(long)max_height;j++)
 @z
 
+@x [65] l.1339
+  sprintf(new_graph->id,"binary(%lu,%lu,%d)",
+@y
+  snprintf(new_graph->id,ID_FIELD_SIZE,"binary(%lu,%lu,%d)",
+@z
+
+@x [66] l.1349
+{@+register float ss;
+@y
+{@+float ss;
+@z
+
 @x [66] l.1356
     for (j=2;j<=max_height;j++) {
 @y
     for (j=2;j<=(long)max_height;j++) {
 @z
 
+@x [67] l.1410
+{@+register long *xtab,*ytab,*ltab,*stab;
+@y
+{@+long *xtab,*ytab,*ltab,*stab;
+@z
+
 @x [67] l.1413
   if (ltab[0]>n) {
 @y
   if (ltab[0]>(long)n) {
 @z
 
+@x [71] l.1478
+{@+register char *p=buffer; /* string pointer */
+@y
+{@+char *p=buffer; /* string pointer */
+@z
+
+@x [72] l.1498
+    {@+register char *p=buffer; /* string pointer */
+      register Vertex *u;
+@y
+    {@+char *p=buffer; /* string pointer */
+      Vertex *u;
+@z
+
 @x [74] l.1546
 Graph *complement(g,copy,self,directed)
   Graph *g; /* graph to be complemented */
@@ -331,13 +531,37 @@ Graph *complement(
   long directed) /* should the graph be directed? */
 @z
 
-@x [74] l.1555
+@x [74] l.1552
+  register long n;
+  register Vertex *u;
+  register siz_t delta; /* difference in memory addresses */
   if (g==NULL) panic(missing_operand); /* where's |g|? */
 @y
+  long n;
+  Vertex *u;
+  siz_t delta; /* difference in memory addresses */
   (void) s; (void) d; (void) k; (void) j; (void) i;
   if (g==NULL) panic(missing_operand); /* where's |g|? */
 @z
 
+@x [74] l.1557
+  sprintf(buffer,",%d,%d,%d)",copy?1:0,self?1:0,directed?1:0);
+@y
+  snprintf(buffer,BUF_SIZE,",%d,%d,%d)",copy?1:0,self?1:0,directed?1:0);
+@z
+
+@x [76] l.1592
+for (v=g->vertices;v<g->vertices+n;v++) {@+register Vertex *vv;
+@y
+for (v=g->vertices;v<g->vertices+n;v++) {@+Vertex *vv;
+@z
+
+@x [76] l.1595
+  {@+register Arc *a;
+@y
+  {@+Arc *a;
+@z
+
 @x [78] l.1643
 Graph *gunion(g,gg,multi,directed)
   Graph *g,*gg; /* graphs to be united */
@@ -350,15 +574,45 @@ Graph *gunion(
   long directed) /* should the graph be directed? */
 @z
 
-@x [78] l.1651
+@x [78] l.1648
+  register long n;
+  register Vertex *u;
+  register siz_t delta,ddelta; /* differences in memory addresses */
   if (g==NULL || gg==NULL) panic(missing_operand);
     /* where are |g| and |gg|? */
 @y
+  long n;
+  Vertex *u;
+  siz_t delta,ddelta; /* differences in memory addresses */
   (void) s; (void) d; (void) k; (void) j; (void) i;
   if (g==NULL || gg==NULL) panic(missing_operand);
     /* where are |g| and |gg|? */
 @z
 
+@x [78] l.1654
+  sprintf(buffer,",%d,%d)",multi?1:0,directed?1:0);
+@y
+  snprintf(buffer,BUF_SIZE,",%d,%d)",multi?1:0,directed?1:0);
+@z
+
+@x [79] l.1668
+  register Arc *a;
+  register Vertex *vv=vert_offset(v,delta);
+      /* vertex in |new_graph| corresponding to |v| in |g| */
+  register Vertex *vvv=vert_offset(vv,-ddelta);
+@y
+  Arc *a;
+  Vertex *vv=vert_offset(v,delta);
+      /* vertex in |new_graph| corresponding to |v| in |g| */
+  Vertex *vvv=vert_offset(vv,-ddelta);
+@z
+
+@x [80] l.1702
+{@+register Arc *b;
+@y
+{@+Arc *b;
+@z
+
 @x [81] l.1724
 Graph *intersection(g,gg,multi,directed)
   Graph *g,*gg; /* graphs to be intersected */
@@ -371,13 +625,43 @@ Graph *intersection(
   long directed) /* should the graph be directed? */
 @z
 
-@x [81] l.1732
+@x [81] l.1729
+  register long n;
+  register Vertex *u;
+  register siz_t delta,ddelta; /* differences in memory addresses */
   if (g==NULL || gg==NULL) panic(missing_operand); /* where are |g| and |gg|? */
 @y
+  long n;
+  Vertex *u;
+  siz_t delta,ddelta; /* differences in memory addresses */
   (void) s; (void) d; (void) k; (void) j; (void) i;
   if (g==NULL || gg==NULL) panic(missing_operand); /* where are |g| and |gg|? */
 @z
 
+@x [81] l.1734
+  sprintf(buffer,",%d,%d)",multi?1:0,directed?1:0);
+@y
+  snprintf(buffer,BUF_SIZE,",%d,%d)",multi?1:0,directed?1:0);
+@z
+
+@x [82] l.1752
+for (v=g->vertices;v<g->vertices+n;v++) {@+register Arc *a;
+  register Vertex *vv=vert_offset(v,delta);
+      /* vertex in |new_graph| corresponding to |v| in |g| */
+  register Vertex *vvv=vert_offset(vv,-ddelta);
+@y
+for (v=g->vertices;v<g->vertices+n;v++) {@+Arc *a;
+  Vertex *vv=vert_offset(v,delta);
+      /* vertex in |new_graph| corresponding to |v| in |g| */
+  Vertex *vvv=vert_offset(vv,-ddelta);
+@z
+
+@x [84] l.1790
+{@+register Arc *b=u->tlen; /* previous arc or edge from |vv| to |u| */
+@y
+{@+Arc *b=u->tlen; /* previous arc or edge from |vv| to |u| */
+@z
+
 @x [87] l.1837
 Graph *lines(g,directed)
   Graph *g; /* graph whose lines will become vertices */
@@ -388,13 +672,33 @@ Graph *lines(
   long directed) /* should the graph be directed? */
 @z
 
-@x [87] l.1843
+@x [87] l.1841
+  register long m; /* the number of lines */
+  register Vertex *u;
   if (g==NULL) panic(missing_operand); /* where is |g|? */
 @y
+  long m; /* the number of lines */
+  Vertex *u;
   (void) s; (void) d; (void) k; (void) j; (void) i;
   if (g==NULL) panic(missing_operand); /* where is |g|? */
 @z
 
+@x [89] l.1898
+for (v=g->vertices+g->n-1;v>=g->vertices;v--) {@+register Arc *a;
+  register long mapped=0; /* has |v->map| been set? */
+  for (a=v->arcs;a;a=a->next) {@+register Vertex *vv=a->tip;
+@y
+for (v=g->vertices+g->n-1;v>=g->vertices;v--) {@+Arc *a;
+  long mapped=0; /* has |v->map| been set? */
+  for (a=v->arcs;a;a=a->next) {@+Vertex *vv=a->tip;
+@z
+
+@x [91] l.1939
+sprintf(buffer,"%.*s-%c%.*s",(BUF_SIZE-3)/2,v->name,@|
+@y
+snprintf(buffer,BUF_SIZE,"%.*s-%c%.*s",(BUF_SIZE-3)/2,v->name,@|
+@z
+
 @x [92] l.1948
     do@+{gb_new_arc(u,v,1L);
 @y
@@ -407,6 +711,14 @@ Graph *lines(
     } while (v->u.V==u->v.V);
 @z
 
+@x [93] l.1963
+for (u=new_graph->vertices;u<new_graph->vertices+m;u++) {@+register Vertex *vv;
+  register Arc *a;@+register long mapped=0;
+@y
+for (u=new_graph->vertices;u<new_graph->vertices+m;u++) {@+Vertex *vv;
+  Arc *a;@+long mapped=0;
+@z
+
 @x [95] l.2011
 Graph *product(g,gg,type,directed)
   Graph *g,*gg; /* graphs to be multiplied */
@@ -419,13 +731,51 @@ Graph *product(
   long directed) /* should the graph be directed? */
 @z
 
-@x [95] l.2018
+@x [95] l.2016
+  register Vertex *u,*vv;
+  register long n; /* the number of vertices in the product graph */
   if (g==NULL || gg==NULL) panic(missing_operand); /* where are |g| and |gg|? */
 @y
+  Vertex *u,*vv;
+  long n; /* the number of vertices in the product graph */
   (void) s; (void) d; (void) k; (void) j; (void) i;
   if (g==NULL || gg==NULL) panic(missing_operand); /* where are |g| and |gg|? */
 @z
 
+@x [96] l.2047
+  sprintf(buffer,"%.*s,%.*s",BUF_SIZE/2-1,v->name,(BUF_SIZE-1)/2,vv->name);
+@y
+  snprintf(buffer,BUF_SIZE,"%.*s,%.*s",BUF_SIZE/2-1,v->name,(BUF_SIZE-1)/2,vv->name);
+@z
+
+@x [96] l.2051
+sprintf(buffer,",%d,%d)",(type?2:0)-(int)(type&1),directed?1:0);
+@y
+snprintf(buffer,BUF_SIZE,",%d,%d)",(type?2:0)-(int)(type&1),directed?1:0);
+@z
+
+@x [97] l.2055
+{@+register Vertex *uu,*uuu;
+  register Arc *a;
+  register siz_t delta; /* difference in memory addresses */
+@y
+{@+Vertex *uu,*uuu;
+  Arc *a;
+  siz_t delta; /* difference in memory addresses */
+@z
+
+@x [99] l.2094
+  register siz_t delta,ddelta;
+@y
+  siz_t delta,ddelta;
+@z
+
+@x [99] l.2103
+      for (u=gg->vertices;u<gg->vertices+gg->n;u++) {@+register Arc *aa;
+@y
+      for (u=gg->vertices;u<gg->vertices+gg->n;u++) {@+Arc *aa;
+@z
+
 @x [100] l.2161
 @d ind z.I
 
@@ -444,6 +794,12 @@ Graph *bi_complete(
   long directed) /* should all arcs go from first part to second? */
 @z
 
+@x [101] l.2181
+      sprintf(new_graph->id,"bi_complete(%lu,%lu,%d)",@|
+@y
+      snprintf(new_graph->id,ID_FIELD_SIZE,"bi_complete(%lu,%lu,%d)",@|
+@z
+
 @x [102] l.2207
 vertices of~$g$ has $m'n+mn'^2$ arcs.
 
@@ -466,6 +822,12 @@ Graph *wheel(
   long directed) /* should all arcs go from center to rim and around? */
 @z
 
+@x [103] l.2237
+      sprintf(new_graph->id,"wheel(%lu,%lu,%d)",@|
+@y
+      snprintf(new_graph->id,ID_FIELD_SIZE,"wheel(%lu,%lu,%d)",@|
+@z
+
 @x [104] l.2245
 extern Graph *bi_complete();
 extern Graph *wheel(); /* standard applications of |induced| */
@@ -492,9 +854,15 @@ Graph *induced(
   long directed) /* should the graph be directed? */
 @z
 
-@x [105] l.2259
+@x [105] l.2256
+  register Vertex *u;
+  register long n=0; /* total number of vertices in induced graph */
+  register long nn=0; /* number of negative vertices in induced graph */
   if (g==NULL) panic(missing_operand); /* where is |g|? */
 @y
+  Vertex *u;
+  long n=0; /* total number of vertices in induced graph */
+  long nn=0; /* number of negative vertices in induced graph */
   (void) s; (void) d; (void) i;
   if (g==NULL) panic(missing_operand); /* where is |g|? */
 @z
@@ -505,8 +873,66 @@ Graph *induced(
 @<Determine \9{n}|n| and \9{n}|nn|@>;
 @z
 
+@x [106] l.2276
+sprintf(buffer,",%s,%d,%d,%d)",@|description?description:null_string,@|
+@y
+snprintf(buffer,BUF_SIZE,",%s,%d,%d,%d)",@|description?description:null_string,@|
+@z
+
 @x [107] l.2280
 @ @<Determine |n| and |nn|@>=
 @y
 @ @<Determine \9{n}|n| and \9{n}|nn|@>=
 @z
+
+@x [108] l.2310
+  sprintf(buffer,"%ld",-k);
+@y
+  snprintf(buffer,BUF_SIZE,"%ld",-k);
+@z
+
+@x [108] l.2322
+        sprintf(buffer,"%s'",v->name);
+@y
+        snprintf(buffer,BUF_SIZE,"%s'",v->name);
+@z
+
+@x [108] l.2328
+      sprintf(buffer,"%.*s:%ld",BUF_SIZE-12,v->name,j);
+@y
+      snprintf(buffer,BUF_SIZE,"%.*s:%ld",BUF_SIZE-12,v->name,j);
+@z
+
+@x [110] l.2359
+  if (u) {@+register Arc *a;@+register Vertex *uu,*vv;
+@y
+  if (u) {@+Arc *a;@+Vertex *uu,*vv;
+@z
+
+@x [113] l.2411
+{@+register Arc *b=uu->tlen; /* existing arc or edge from |u| to |uu| */
+@y
+{@+Arc *b=uu->tlen; /* existing arc or edge from |u| to |uu| */
+@z
+
+@x [114] l.2423
+{@+register Graph *gg=v->subst;
+  register Vertex *vv=gg->vertices;
+  register Arc *a;
+@y
+{@+Graph *gg=v->subst;
+  Vertex *vv=gg->vertices;
+  Arc *a;
+@z
+
+@x [114] l.2428
+    sprintf(buffer,"%.*s:%.*s",BUF_SIZE/2-1,v->name,(BUF_SIZE-1)/2,vv->name);
+@y
+    snprintf(buffer,BUF_SIZE,"%.*s:%.*s",BUF_SIZE/2-1,v->name,(BUF_SIZE-1)/2,vv->name);
+@z
+
+@x [114] l.2430
+    for (a=vv->arcs;a;a=a->next) {@+register Vertex *vvv=a->tip;
+@y
+    for (a=vv->arcs;a;a=a->next) {@+Vertex *vvv=a->tip;
+@z
diff --git a/PROTOTYPES/gb_flip.ch b/PROTOTYPES/gb_flip.ch
index 03b9d00..88654f8 100644
--- a/PROTOTYPES/gb_flip.ch
+++ b/PROTOTYPES/gb_flip.ch
@@ -49,16 +49,22 @@ extern long gb_flip_cycle(void); /* compute 55 more pseudo-random numbers */
 
 @x [7] l.134
 long gb_flip_cycle()
+{@+register long *ii, *jj;
 @y
 long gb_flip_cycle(void)
+{@+long *ii, *jj;
 @z
 
 @x [8] l.159
 void gb_init_rand(seed)
     long seed;
+{@+register long i;
+  register long prev=seed, next=1;
 @y
 void gb_init_rand(
     long seed)
+{@+long i;
+  long prev=seed, next=1;
 @z
 
 @x [10] l.223
@@ -82,9 +88,13 @@ extern void gb_init_rand(long);
 @x [12] l.252
 long gb_unif_rand(m)
     long m;
+{@+register unsigned long t=two_to_the_31-(two_to_the_31 % m);
+  register long r;
 @y
 long gb_unif_rand(
     long m)
+{@+unsigned long t=two_to_the_31-(two_to_the_31 % m);
+  long r;
 @z
 
 @x [12] l.256
diff --git a/PROTOTYPES/gb_graph.ch b/PROTOTYPES/gb_graph.ch
index 220efef..c660d09 100644
--- a/PROTOTYPES/gb_graph.ch
+++ b/PROTOTYPES/gb_graph.ch
@@ -107,15 +107,21 @@ Graph *gb_new_graph(
   long n) /* desired number of vertices */
 @z
 
+@x [23] l.454
+      sprintf(cur_graph->id,"gb_new_graph(%ld)",n);
+@y
+      snprintf(cur_graph->id,ID_FIELD_SIZE,"gb_new_graph(%ld)",n);
+@z
+
 @x [25] l.478
 extern void make_compound_id();
   /* routine to set one |id| field from another */
 extern void make_double_compound_id(); /* ditto, but from two others */
 @y
-extern void make_compound_id(Graph *,char *,Graph *,char *);
+extern void make_compound_id(Graph *,const char *,Graph *,const char *);
   /* routine to set one |id| field from another */
-extern void make_double_compound_id(Graph *,char *,Graph *,char *,@|
-  Graph *,char *); /* ditto, but from two others */
+extern void make_double_compound_id(Graph *,const char *,Graph *,const char *,@|
+  Graph *,const char *); /* ditto, but from two others */
 @z
 
 @x [26] l.487
@@ -125,17 +131,19 @@ void make_compound_id(g,s1,gg,s2) /* |sprintf(g->id,"%s%s%s",s1,gg->id,s2)| */
   Graph *gg; /* graph whose |id| is to be copied */
   char *s2; /* string for the end of the new |id| */
 @y
-void make_compound_id( /* |sprintf(g->id,"%s%s%s",s1,gg->id,s2)| */
+void make_compound_id( /* |snprintf(g->id,ID_FIELD_SIZE,"%s%s%s",s1,gg->id,s2)| */
   Graph *g, /* graph whose |id| is to be set */
-  char *s1, /* string for the beginning of the new |id| */
+  const char *s1, /* string for the beginning of the new |id| */
   Graph *gg, /* graph whose |id| is to be copied */
-  char *s2) /* string for the end of the new |id| */
+  const char *s2) /* string for the end of the new |id| */
 @z
 
 @x [26] l.495
   if (strlen(tmp)<avail) sprintf(g->id,"%s%s%s",s1,tmp,s2);
+  else sprintf(g->id,"%s%.*s...)%s",s1,avail-5,tmp,s2);
 @y
-  if (strlen(tmp)<(siz_t)avail) sprintf(g->id,"%s%s%s",s1,tmp,s2);
+  if (strlen(tmp)<(siz_t)avail) snprintf(g->id,ID_FIELD_SIZE,"%s%s%s",s1,tmp,s2);
+  else snprintf(g->id,ID_FIELD_SIZE,"%s%.*s...)%s",s1,avail-5,tmp,s2);
 @z
 
 @x [27] l.500
@@ -149,18 +157,19 @@ void make_double_compound_id(g,s1,gg,s2,ggg,s3)
   char *s3; /* string for the end of the new |id| */
 @y
 void make_double_compound_id(
-    /* |sprintf(g->id,"%s%s%s%s%s",s1,gg->id,s2,ggg->id,s3)| */
+    /* |snprintf(g->id,ID_FIELD_SIZE,"%s%s%s%s%s",s1,gg->id,s2,ggg->id,s3)| */
   Graph *g, /* graph whose |id| is to be set */
-  char *s1, /* string for the beginning of the new |id| */
+  const char *s1, /* string for the beginning of the new |id| */
   Graph *gg, /* first graph whose |id| is to be copied */
-  char *s2, /* string for the middle of the new |id| */
+  const char *s2, /* string for the middle of the new |id| */
   Graph *ggg, /* second graph whose |id| is to be copied */
-  char *s3) /* string for the end of the new |id| */
+  const char *s3) /* string for the end of the new |id| */
 @z
 
 @x [27] l.509
   if (strlen(gg->id)+strlen(ggg->id)<avail)
     sprintf(g->id,"%s%s%s%s%s",s1,gg->id,s2,ggg->id,s3);
+  else sprintf(g->id,"%s%.*s...)%s%.*s...)%s",s1,avail/2-5,gg->id,
 @y
   if (strlen(gg->id)+strlen(ggg->id)<(siz_t)avail)
   { int ret = snprintf(g->id,ID_FIELD_SIZE,"%s%s%s%s%s",s1,gg->id,s2,ggg->id,s3);
@@ -169,22 +178,27 @@ void make_double_compound_id(
       abort();
     }
   }
+  else snprintf(g->id,ID_FIELD_SIZE,"%s%.*s...)%s%.*s...)%s",s1,avail/2-5,gg->id,
 @z
 
 @x [29] l.551
 Arc *gb_virgin_arc()
+{@+register Arc *cur_arc=next_arc;
 @y
 Arc *gb_virgin_arc(void)
+{@+Arc *cur_arc=next_arc;
 @z
 
 @x [30] l.583
 void gb_new_arc(u,v,len)
   Vertex *u, *v; /* a newly created arc will go from |u| to |v| */
   long len; /* its length */
+{@+register Arc *cur_arc=gb_virgin_arc();
 @y
 void gb_new_arc(
   Vertex *u,Vertex *v, /* a newly created arc will go from |u| to |v| */
   long len) /* its length */
+{@+Arc *cur_arc=gb_virgin_arc();
 @z
 
 @x [31] l.623
@@ -199,18 +213,24 @@ void gb_new_arc(
 void gb_new_edge(u,v,len)
   Vertex *u, *v; /* new arcs will go from |u| to |v| and from |v| to |u| */
   long len; /* their length */
+{@+register Arc *cur_arc=gb_virgin_arc();
 @y
 void gb_new_edge(
   Vertex *u,Vertex *v, /* new arcs will go from |u| to |v| and from |v| to |u| */
   long len) /* their length */
+{@+Arc *cur_arc=gb_virgin_arc();
 @z
 
 @x [35] l.691
 char *gb_save_string(s)
   register char *s; /* the string to be copied */
+{@+register char *p=s;
+  register long len;
 @y
 char *gb_save_string(
-  register char *s) /* the string to be copied */
+  const char *s) /* the string to be copied */
+{@+char *p=(char*)s;
+  long len;
 @z
 
 @x [39] l.774
@@ -246,7 +266,7 @@ extern Arc*gb_virgin_arc(void);
    /* allocate a new |Arc| record */
 extern void gb_new_edge(Vertex *,Vertex *,long);
    /* append an edge (two arcs) to the current graph */
-extern char*gb_save_string(register char *);
+extern char*gb_save_string(const char *);
    /* store a string in the current graph */
 extern void switch_to_graph(Graph *);
    /* save allocation variables, swap in others */
@@ -275,17 +295,31 @@ extern Vertex* hash_lookup(char *,Graph *);
 @x [44] l.857
 void hash_in(v)
   Vertex *v;
+{@+  register char *t=v->name;
+  register Vertex *u;
 @y
 void hash_in(
   Vertex *v)
+{@+char *t=v->name;
+  Vertex *u;
+@z
+
+@x [45] l.885
+{@+register long h;
+@y
+{@+long h;
 @z
 
 @x [46] l.900
 Vertex* hash_out(s)
   char* s;
+{@+register char *t=s;
+  register Vertex *u;
 @y
 Vertex* hash_out(
   char *s)
+{@+char *t=s;
+  Vertex *u;
 @z
 
 @x [47] l.911
@@ -296,6 +330,12 @@ void hash_setup(
   Graph *g)
 @z
 
+@x [47] l.914
+  if (g && g->n>0) {@+register Vertex *v;
+@y
+  if (g && g->n>0) {@+Vertex *v;
+@z
+
 @x [48] l.926
 Vertex* hash_lookup(s,g)
   char *s;
@@ -305,3 +345,9 @@ Vertex* hash_lookup(
   char *s,
   Graph *g)
 @z
+
+@x [48] l.930
+  if (g && g->n>0) {@+register Vertex *v;
+@y
+  if (g && g->n>0) {@+Vertex *v;
+@z
diff --git a/PROTOTYPES/gb_io.ch b/PROTOTYPES/gb_io.ch
index 705cb29..da5513e 100644
--- a/PROTOTYPES/gb_io.ch
+++ b/PROTOTYPES/gb_io.ch
@@ -52,8 +52,10 @@ int main(void)
 
 @x [9] l.123
 static void fill_buf()
+{@+register char *p;
 @y
 static void fill_buf(void)
+{@+char *p;
 @z
 
 @x [11] l.182
@@ -66,7 +68,7 @@ static void fill_buf(void)
 static char *imap="0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ\
 abcdefghijklmnopqrstuvwxyz_^~&@@,;.:?!%#$+-*/|\\<=>()[]{}`'\" \n";
 @y
-static char *imap=
+static const char *imap=
    "0123456789"@|
    "ABCDEFGHIJKLMNOPQRSTUVWXYZ"@|
    "abcdefghijklmnopqrstuvwxyz"@|
@@ -113,14 +115,18 @@ extern long imap_ord(char); /* the ordinal number of a given character */
 
 @x [15] l.213
 static void icode_setup()
+{@+register long k;
+  register char *p;
 @y
 static void icode_setup(void)
+{@+long k;
+  char *p;
 @z
 
 @x [15] l.217
   for (p=imap,k=0; *p; p++,k++) icode[*p]=k;
 @y
-  for (p=imap,k=0; *p; p++,k++) icode[(unsigned char)*p]=k;
+  for (p=(char*)imap,k=0; *p; p++,k++) icode[(unsigned char)*p]=k;
 @z
 
 @x [16] l.225
@@ -135,10 +141,14 @@ extern long new_checksum(char *,long); /* compute change in magic number */
 long new_checksum(s,old_checksum)
   char *s; /* a string */
   long old_checksum;
+{@+register long a=old_checksum;
+  register char*p;
 @y
 long new_checksum(
   char *s, /* a string */
   long old_checksum)
+{@+long a=old_checksum;
+  char*p;
 @z
 
 @x [18] l.245
@@ -204,9 +214,11 @@ long gb_digit(
 @x [24] l.323
 unsigned long gb_number(d)
     char d;
+{@+register unsigned long a=0;
 @y
 unsigned long gb_number(
     char d)
+{@+unsigned long a=0;
 @z
 
 @x [24] l.328
@@ -258,8 +270,8 @@ char *gb_string(
 extern void gb_raw_open(); /* open a file for GraphBase input */
 extern long gb_open(); /* open a GraphBase data file; return 0 if OK */
 @y
-extern void gb_raw_open(char *); /* open a file for GraphBase input */
-extern long gb_open(char *); /* open a GraphBase data file; return 0 if OK */
+extern void gb_raw_open(const char *); /* open a file for GraphBase input */
+extern long gb_open(const char *); /* open a GraphBase data file; return 0 if OK */
 @z
 
 @x [30] l.432
@@ -267,7 +279,13 @@ void gb_raw_open(f)
     char *f;
 @y
 void gb_raw_open(
-    char *f)
+    const char *f)
+@z
+
+@x [31] l.458
+  sprintf(str_buf,"%s%s",DATA_DIRECTORY,f);
+@y
+  snprintf(str_buf,STR_BUF_LENGTH,"%s%s",DATA_DIRECTORY,f);
 @z
 
 @x [32] l.464
@@ -275,7 +293,13 @@ long gb_open(f)
     char *f;
 @y
 long gb_open(
-    char *f)
+    const char *f)
+@z
+
+@x [34] l.501
+sprintf(str_buf,"* File \"%s\"",f);
+@y
+snprintf(str_buf,STR_BUF_LENGTH,"* File \"%s\"",f);
 @z
 
 @x [38] l.531
@@ -290,6 +314,12 @@ long gb_close()
 long gb_close(void)
 @z
 
+@x [39] l.541
+  sprintf(str_buf,"* End of file \"%s\"",file_name);
+@y
+  snprintf(str_buf,STR_BUF_LENGTH,"* End of file \"%s\"",file_name);
+@z
+
 @x [41] l.565
 @ @d gb_raw_close gb_r_close /* for Procrustean external linkage */
 
diff --git a/PROTOTYPES/gb_sort.ch b/PROTOTYPES/gb_sort.ch
index 94f341f..dea017b 100644
--- a/PROTOTYPES/gb_sort.ch
+++ b/PROTOTYPES/gb_sort.ch
@@ -36,9 +36,15 @@ extern node* gb_sorted[]; /* the results of |gb_linksort| */
 @x [5] l.95
 void gb_linksort(l)
   node *l;
+{@+register long k; /* index to destination list */
+  register node **pp; /* current place in list of pointers */
+  register node *p, *q; /* pointers for list manipulation */
 @y
 void gb_linksort(
   node *l)
+{@+long k; /* index to destination list */
+  node **pp; /* current place in list of pointers */
+  node *p, *q; /* pointers for list manipulation */
 @z
 
 @x [5] l.101
-- 
2.54.0

